assignment/impl/src/java/org/sakaiproject/assignment/impl/reminder/AssignmentDueReminderServiceImpl.java

```diff
@@ -0,0 +1,246 @@
+        } catch (IdUnusedException | PermissionException e) {
+            log.error("Error scheduling due date reminder email.", assignmentId, e);
+        }
+    }
+
+    public void removeScheduledReminder(String assignmentId) {
+        scheduledInvocationManager.deleteDelayedInvocation("org.sakaiproject.assignment.api.reminder.AssignmentDueReminderService", assignmentId);
+    }
+
+    public void execute(String opaqueContext) {
+        log.debug("AssignmentDueReminderService execute");
+
+        try {
+            Assignment assignment = assignmentService.getAssignment(opaqueContext);
+            Site site = siteService.getSite(assignment.getContext());
+
+            // Do not send reminders if the site is unpublished or softly deleted
+            if (site.isPublished() && !site.isSoftlyDeleted()) {
+                for (Member member : site.getMembers()) {
+                    if (assignmentService.canSubmit(assignment, member.getUserId()) && checkEmailPreference(member)) {
+                        sendEmailReminder(site, assignment, member);
+                    }
+                }
+            }
+        } catch (IdUnusedException | PermissionException e) {
+            log.error(e.getMessage(), e);
+        }
+    }
+
+    private void sendEmailReminder(Site site, Assignment assignment, Member submitter) {
+        log.debug("SendEmailReminder: '" + assignment.getTitle() + "' to " + submitter.getUserDisplayId());
+
+        String assignmentTitle = assignment.getTitle();
+        if (assignment.getTitle().length() > 11) {
+            assignmentTitle = assignment.getTitle().substring(0, 11) + "[...]";
+        }
+
+        String courseName = site.getTitle();
+
+        SimpleDateFormat sdf = new SimpleDateFormat("MMM dd, yyyy 'at' hh:mm aa");
+        Instant dueDate = assignment.getDueDate();
+        String formattedDateDue = sdf.format(dueDate);
+
+        String toStr = getUserEmail(submitter.getUserId());
+        if (StringUtils.isEmpty(toStr)) {
+            return;
+        }
+        String headerToStr = getUserDisplayName(submitter.getUserId()) + " <" + getUserEmail(submitter.getUserId()) + ">";
+        String fromStr = "\"" + courseName + "\" <" + getSetupRequest() + ">";
+
+        Set<Member> instructors = new HashSet<>();
+        for (Member member : site.getMembers()) {
+            if (StringUtils.equals(member.getRole().getId(), site.getMaintainRole())) {
+                instructors.add(member);
+            }
+        }
+        String replyToStr = getReplyTo(instructors);
+        log.debug("Reply to string: " + replyToStr);
+
+        String subject = "Assignment '" + assignmentTitle + "' Due on " + formattedDateDue;
+
+        StringBuilder body = new StringBuilder();
+        body.append("Hello ");
+        body.append(getUserFirstName(submitter.getUserId()));
+        body.append(",<br />");
+        body.append("<br />");
+        body.append("Reminder: An assignment of yours is due within 24 hours.");
```

Comment:
I was about to approve and merge, but just happened to notice these strings here. I think these should be sourced from message bundles to support internationalization.

Reply:
Yes, I believe the original thought was to get this merged with the known internationalization issue so that it's in 19. Then the strings could be addressed with a new Jira before 19.0 is released.

Reply:
Great, thanks @davidpbauer